docs/proposed/psa-driver-interface.md

```diff
@@ -536,6 +537,37 @@ The core will only call `acme_export_public_key` on a private key. Drivers imple
 
 The core guarantees that the size of the output buffer (`data_size`) is sufficient to export any key with the given attributes. The driver must set `*data_length` to the exact size of the exported key.
 
+### Builtin keys for opaque drivers
+
+An opaque driver can be pre-provisioned from the factory with keys. These keys are never seen by the PSA Crypto core, but in order to be able to use these keys, they need to be declared to the core.
+
+An opaque driver can declare such keys in its JSON descriptor file. The declaration is an array of strings, where each string is the name of a preprovisioned key, e.g.:
+```
+{
+    "builtin_keys": ["pre_key_1", "pre_key_2"]
+}
+```
+
+The key names need to be compatible with C language variable naming constraints, i.e. they cannot contain spaces and cannot start with a digit.
+
+For each declared key, the respective driver needs to define and externally declare the corresponding PSA Crypto key attributes (`psa_key_attributes_t`). The name of this variable, for each key, is `{prefix}_builtin_{keyname}_attributes`, where `{prefix}` is the implementing driver's prefix (lowercase), and `{keyname}` is the lowercase name of the key from the `builtin_keys` array.
+
+E.g. for a key called `"identity"`, a driver called `"acme"` would need to expose:
+```
+extern psa_core_key_attributes_t acme_builtin_identity_attributes;
```

Comment:
An attribute structure include a key id and lifetime information. Presumably the key id should be ignored in favor of the one declared as `{PREFIX}_BUILTIN_{KEYNAME}_ID`, and the location should be ignored in favor of the driver's location, but the persistence should be taken into account? Or is it an error if the key id or location are wrong?

Reply:
I can specify that the id should be set to `{PREFIX}_BUILTIN_{KEYNAME}_ID`, and the lifetime set to the combination of the driver's lifetime and the actual key's persistency (which, given it's a preprovisioned one, would be either persistent or persistent-nonerasable).

Reply:
Hmmm. When the key is erasable, its persistence level may need to be configured by the integrator. For example, some deployment processes use a factory-provisioned key throughout the lifetime of the device, while others only use the factory-provisioned key temporarily and change that key when the final customer takes ownership of the device. So maybe there should be an optional override of the persistence level in the JSON?

Reply:
OK, so you're thinking an integrator should be able to tell the core that a preprovisioned key needs to be treated non-eraseable, even though the hardware would support erasing it?

If that's the case, why not define all of the key attributes in JSON? There's more things like this: e.g. a preprovisioned AES key that you should be able to use for any type of block-cipher algorithm. Or a preprovisioned P256 key that could be used for ECDH or ECDSA depending on the integrator.